File: dart/test/client_xlang_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';
import 'package:multipath/multipath.dart';
import 'package:test/test.dart';

// Cross-language proof for the Dart client: the Client against a real JVM Origin over WebSocket
// links, driving a tunnel (echo) and a normal HTTP exchange (greeter). Skipped unless MP_JVM_CP set.
final cp = Platform.environment['MP_JVM_CP'];

Future<(int, Process)> startOrigin(int n) async {
  final proc = await Process.start(
      'java', ['-cp', cp!, 'app.microteams.multipath.OriginMain', '$n']);
  final port = Completer<int>();
  proc.stdout
      .transform(utf8.decoder)
      .transform(const LineSplitter())
      .listen((line) {
    final m = RegExp(r'^LISTENING (\d+)').firstMatch(line);
    if (m != null && !port.isCompleted) port.complete(int.parse(m.group(1)!));
  });
  proc.stderr.transform(utf8.decoder).listen(stderr.write);
  return (await port.future.timeout(const Duration(seconds: 30)), proc);
}

void main() {
  test('tunnels and round-trips HTTP through a JVM origin over WebSocket links',
      () async {
    final n = 3;
    final (port, proc) = await startOrigin(n);
    try {
      final lines = List.generate(n, (_) => 'ws://127.0.0.1:$port');
      final linkEvents = <LinkState>[];
      final client = await Client.dial(
        lines,
        pingInterval: const Duration(milliseconds: 20),
        deadAfter: const Duration(seconds: 2),
        ackInterval: const Duration(milliseconds: 8),
        reconnectDelay: const Duration(milliseconds: 20),
        maxDelay: const Duration(milliseconds: 200),
        onLinkState: linkEvents.add,
      );

      // Tunnel: write, half-close, read the echo back to EOF.
      final st = client.openTunnel('echo:0', ticket: utf8.encode('ticket'));
      final msg = Uint8List.fromList(
          List.generate(20 * 1024, (i) => (i * 7 + 3) & 0xff));
      await st.write(msg);
      st.closeWrite();
      final echoed = BytesBuilder();
      while (true) {
        final chunk = await st.read();
        if (chunk == null) break;
        echoed.add(chunk);
      }
      expect(echoed.takeBytes(), msg);

      // Normal: an HTTP round trip via the client's roundTrip helper — the library serializes the
      // request and parses the response (status/headers/body), so the caller writes no HTTP by hand.
      final resp = await client
          .roundTrip(MultipathRequest('GET', Uri.parse('http://origin/xlang')));
      expect(resp.statusCode, 200);
      expect(utf8.decode(resp.body), endsWith('hello /xlang'));

      // The redundant transport exposes per-line health, and up transitions were reported.
      final stats = client.stats();
      expect(stats.length, n);
      expect(stats.every((s) => s.state == 'up'), isTrue);
      expect(linkEvents.any((e) => e.up), isTrue);

      client.close();
    } finally {
      proc.kill();
    }
  },
      timeout: const Timeout(Duration(seconds: 40)),
      skip: cp == null ? 'set MP_JVM_CP' : false);
}

File: ts/test/client.xlang.test.ts
import { spawn } from 'node:child_process';
import { createInterface } from 'node:readline';
import { describe, expect, it } from 'vitest';
import { Client } from '../src/client.js';

// Cross-language proof for the whole browser client: the TypeScript Client (global WebSocket standing
// in for the browser's) against a real JVM Origin, driving both routed paths — a tunnel spliced to an
// echo and a normal HTTP exchange spliced to a greeter. Skipped unless MP_JVM_CP is set.
const cp = process.env.MP_JVM_CP;
const maybe = cp ? it : it.skip;

async function startOrigin(n: number): Promise<{ port: number; kill: () => void }> {
  const proc = spawn('java', ['-cp', cp!, 'app.microteams.multipath.OriginMain', String(n)]);
  proc.stderr.pipe(process.stderr);
  const rl = createInterface({ input: proc.stdout });
  const port = await new Promise<number>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error('java origin did not report LISTENING')),
      30000,
    );
    rl.on('line', (line) => {
      const m = /^LISTENING (\d+)/.exec(line);
      if (m) {
        clearTimeout(timer);
        resolve(Number(m[1]));
      }
    });
  });
  return { port, kill: () => proc.kill() };
}

const fast = {
  pingIntervalMs: 20,
  deadAfterMs: 2000,
  ackIntervalMs: 8,
  reconnectDelayMs: 20,
  maxDelayMs: 200,
};

describe('browser client cross-language', () => {
  maybe(
    'tunnels and round-trips HTTP through a JVM origin over WebSocket links',
    async () => {
      const n = 3;
      const origin = await startOrigin(n);
      try {
        const lines = Array.from({ length: n }, () => `ws://127.0.0.1:${origin.port}`);
        const client = await Client.dial(lines, fast);

        // Tunnel: write, half-close, read the echo back to EOF.
        const st = client.open('echo', new TextEncoder().encode('ticket'));
        const msg = new Uint8Array(20 * 1024).map((_, i) => (i * 7 + 3) & 0xff);
        await st.write(msg);
        st.closeWrite();
        const echoed: number[] = [];
        for (;;) {
          const chunk = await st.read();
          if (chunk === null) break;
          echoed.push(...chunk);
        }
        expect(Uint8Array.from(echoed)).toEqual(msg);

        // Normal: an HTTP round trip to the origin's own greeter.
        const resp = await client.fetch('greeter', new Request('http://origin/xlang'));
        expect(resp.status).toBe(200);
        expect(await resp.text()).toBe('hello /xlang');

        // WebSocket: a real RFC 6455 handshake and message round trip through the substrate to the
        // origin's ws-echo backend — proves the browser client's application-level WebSocket, not
        // just a byte tunnel. Binary only: the origin's test echo reuses the link layer's
        // byte-stream WebSocketConn, which (by original design — see websocket.go's own header
        // comment) only understands OP_BINARY; a text frame throws there and drops the connection.
        const ws = await client.openWebSocket('ws-echo', '/echo');
        const gotMessages: ArrayBuffer[] = [];
        ws.onmessage = (ev) => gotMessages.push(ev.data as ArrayBuffer);
        ws.send(new TextEncoder().encode('hello ws'));
        ws.send(new TextEncoder().encode('second message'));
        await new Promise((resolve) => setTimeout(resolve, 500));
        expect(gotMessages.length).toBe(2);
        expect(new TextDecoder().decode(gotMessages[0])).toBe('hello ws');
        expect(new TextDecoder().decode(gotMessages[1])).toBe('second message');
        ws.close();

        client.close();
      } finally {
        origin.kill();
      }
    },
    40000,
  );
});
